Docs for CCCL configuration macros (#7280)
NVIDIA/cccl: docs/cccl/config_macros.rst

.. code-block:: bash
    nvcc -DCCCL_DISABLE_SOME_FEATURE src.cu

.. important::
    These macros should be defined consistently in the whole project. Defining them only in a single module or compilation phase may lead to unexpected compile time and runtime errors.

+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_ARCH_DEPENDENT_NAMESPACE             | ???                                                                                                                +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_BF16_SUPPORT                         | Disables support for ``__nv_bfloat16`` type.                                                                       +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+

> | CCCL_DISABLE_CDP                                  | Disables CUDA Dynamic Parallelism.                                                                                 +

Reviewer: Suggestion: Maybe we should phrase the descriptions in a way that rules out that the CUDA language/runtime feature is disabled. "Disables CUDA Dynamic Parallelism." could be interpreted as it disables support for calling kernels from other device functions. But it just disables CCCL's use of that feature. So maybe:

```suggestion
| CCCL_DISABLE_CDP                                  | Disables use of CUDA Dynamic Parallelism.                                                                                 +
```
Consider reviewing the other descriptions with that in mind.

Author: That's true, I was considering this, but thought - it's called CCCL configuration macros, so it should be obvious that's it's controlling only CCCL features.

Anyway, you are right, I will fix this, thanks for the suggestion

+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+
| CCCL_ENABLE_DEVICE_ASSERTIONS                     | Enables assertions in device code exclusively.                                                                     +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+
| CCCL_ENABLE_EXPERIMENTAL_HOST_ATOMICS_128B        | Enables experimental support for 128b atomics in host code.                                                        +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+
| CCCL_ENABLE_HOST_ASSERTIONS                       | Enables assertions in host code exclusively.                                                                       +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------+

> | CCCL_ENABLE_OPTIONAL_REF                          | Enables ``optional<T&>`` as defined in P2988R8.                                                                    +

Reviewer: @miscco should we just turn `optional<T&>` on by default and remove this macro?

Author: We still need to update the implementation from R8 to R12, I have a branch where I started implementing it already

CCCL configuration macros
=========================

The CUDA Core Compute Libraries provide set of macros to enable or disable specific features. These macros must be defined before any CCCL source file is included. The recommended way is to define them as the predefined compiler macros, for example:

.. code-block:: bash
    nvcc -DCCCL_DISABLE_SOME_FEATURE src.cu

.. important::
    These macros should be defined consistently in the whole project. Defining them only in a single module or compilation phase may lead to unexpected compile time and runtime errors.

+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_ARCH_DEPENDENT_NAMESPACE             | ???                                                                                                                                                    +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

> | CCCL_DISABLE_BF16_SUPPORT                         | Disables use and library support for ``__nv_bfloat16`` type.                                                                                           +

Reviewer 2: We should move the note below into this and the FP16 description
```suggestion
| CCCL_DISABLE_BF16_SUPPORT                         | Disables use and library support for ``__nv_bfloat16``. Also disables support for smaller floating point types.                                                                                           +
```

+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_DEVICE_RUNTIME                       | Disables use of CUDA device runtime APIs (``<cuda_device_runtime.h>``), thus makes some APIs that are ``__host__ __device__`` to be ``__host__`` only. +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_DLPACK                               | Disables inclusion of DLPack header and APIs.                                                                                                          +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

> | CCCL_DISABLE_EXCEPTIONS                           | Disables throwing exceptions. Each ``throw`` is replaced by ``cuda::std::terminate()`` call.                                                           +

Reviewer 2: ```suggestion
| CCCL_DISABLE_EXCEPTIONS                           | Disables throwing exceptions. Each ``throw`` is replaced by a ``cuda::std::terminate()`` call.                                                           +
```

+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_RTTI                                 | Disables use of for runtime type information.                                                                                                          +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_ENABLE_ASSERTIONS                            | Enables assertions in both host and device code.                                                                                                       +
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

> | CCCL_ENABLE_DEVICE_ASSERTIONS                     | Enables assertions in device code exclusively.                                                                                                         +

Reviewer 2: ```suggestion
| CCCL_ENABLE_DEVICE_ASSERTIONS                     | Enables assertions in device code, independent of debug mode.                                                                                                         +
```

| CCCL_ENABLE_HOST_ASSERTIONS                       | Enables assertions in host code, independent of debug mode.                                                                                            |
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

C++ Feature Control Macros
--------------------------
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_EXCEPTIONS                           | Disables throwing exceptions. Each ``throw`` is replaced by a ``cuda::std::terminate()`` call.                                                         |
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

> | CCCL_DISABLE_RTTI                                 | Disables use of for runtime type information.                                                                                                          |

Reviewer: ```suggestion
| CCCL_DISABLE_RTTI                                 | Disables use of runtime type information.                                                                                                          |
```

CUDA Feature Control Macros
---------------------------
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_ARCH_DEPENDENT_NAMESPACE             | Disables architecture dependent name mangling of kernels.                                                                                              |
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+
| CCCL_DISABLE_CDP                                  | Disables use of CUDA Dynamic Parallelism.                                                                                                              |
+---------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+

> | CCCL_DISABLE_CTK_COMPATIBILITY_CHECK              | Disables the check whether NVCC version matches the CUDA Toolkit version.                                                                              |

Reviewer: ```suggestion
| CCCL_DISABLE_CTK_COMPATIBILITY_CHECK              | Disables the check whether NVCC's version matches the CUDA Toolkit version.                                                                              |
```
